lab_3_ann_retriever/main.py

```diff
@@ -426,6 +549,63 @@ def build(self, vectors: list[Vector]) -> bool:
 
         In case of corrupt input arguments, False is returned.
         """
+        if not isinstance(vectors, list) or not vectors \
+                or not all(isinstance(vector, (tuple, list)) for vector in vectors):
+            return False
+
+        space = [
+            {
+             'vectors': [(vector, index) for index, vector in enumerate(vectors)],
+             'depth': 0,
+             'parent node': Node(),
+             'current dimension': True
+            }
+        ]
+        dimensions = len(vectors[0])
+
+        while space:
+            current_vectors = space[0]['vectors']
+            depth = space[0]['depth']
+            parent_node = space[0]['parent node']
+            current_dimension = space.pop(0)['current dimension']
+            if not isinstance(current_vectors, list) or not isinstance(depth, int):
+                return False
+            if not current_vectors:
+                continue
+            axis = depth % dimensions
+            current_vectors.sort(key=lambda vector: float(vector[0][axis]))
+            median_index = len(current_vectors) // 2
+            median_node = Node(current_vectors[median_index][0], current_vectors[median_index][1])
+
+            if parent_node.payload == -1:
+                self._root = median_node
+            else:
+                if not current_dimension:
+                    parent_node.right_node = median_node
+                else:
+                    parent_node.left_node = median_node
+
+            space.append(
+                {
+                    'vectors': current_vectors[:median_index],
+                    'depth': depth + 1,
+                    'parent node': median_node,
+                    'current dimension': True
+                }
+            )
+            space.append(
+                {
+                    'vectors': current_vectors[median_index + 1:],
+                    'depth': depth + 1,
+                    'parent node': median_node,
+                    'current dimension': False
+                }
+            )
+
+        if not self._root:
```

Comment:
as I understand it, you do this after the monstrous construction of an entire tree?